### DISCUSSION TITLE
Proposial of "OppositeJoint" for object (pipes, profiles, ..) chaining without dealing with polarity.

### DISCUSSION THREAD

**Discussion by roman-dvorak** [gumyr/build123d#637] [Ideas]
Hi,

I would like to propose a new type of joint. While experimenting with Build123d, I noticed an issue: when I want to connect multiple objects in sequence, I have to deal with the "output" and "input" of each joint separately. This can be seen in the [Build123d examples](https://build123d.readthedocs.io/en/latest/joints.html#rigid-joint), where "inlet" and "outlet" are defined, with one axis pointing outwards and the other inwards. This at least complicates the way objects are chained together because I have to manage the "polarity" of the joint. Consequently, it's not possible to connect objects in any arbitrary manner.

I propose a concept I call `OppositeJoint`, which would connect vertices that face "against each other." For example, with pipes from the examples or my aluminum profiles, these could be chained without any restrictions or concerns about polarity. Additionally, they could be connected on both sides to a linear joint.

Thank you for considering my suggestion.

**Comment by gumyr** [comment]
I need more information to understand your suggestion.  All objects are placed with a `Location` which defines both the position and orientation (i.e. all 6 degrees of freedom).  To "flip" a `Location` (i.e. the orientation is reversed - say "up" to "down") one can use the `-` operator so `-joint_location` is the opposite of `joint_location`.  Is `OppositeJoint` an alternative to flipping the location?

**Comment by roman-dvorak** [comment]
Thanks for your response, and sorry for my late reply. I might need to illustrate my request a bit.

The current implementation of `RigidJoints` is great for connections where orientation (polarity) matters. A real-world example is quick-release couplings for pneumatics (see the [link](https://www.amazon.com/PJRYC-Pneumatic-European-Connector-Compressor/dp/B09Q17X7RG)).

To explain my concern: I wanted to create a library for generating modular aluminum profiles in my fork of `bd_warehouse` (I want to make PR later) module. These profiles can form very complex structures and I wanted to use build123d features for that. (Therefore this PR #138  would be also useful)

![obrazek](https://github.com/gumyr/build123d/assets/5196729/ebb04b7a-bfcf-4908-960e-f53bf190d5c6)
![obrazek](https://github.com/gumyr/build123d/assets/5196729/9c3d406d-d83f-4604-ad87-47ab07e62bcc)


I encountered the problem of chaining these profiles because orientation of Z-axis matters and it is not possible to stack objects face-to-face. 

My idea was to have build123d `joints` implemented directly in the class generating these profiles.. face-to-face joint at each end. And linear joints along slots for profile nuts. 

![obrazek](https://github.com/gumyr/build123d/assets/5196729/84095957-7555-4fbf-9550-8fdb171c0d19)

The `bd_warehouse` module contains a flange pipes generator, which faces the same issue (I think). I want to avoid dealing with input or output distinctions. For instance, with a T-shaped pipe, it's unclear if there is one or two inputs.

![obrazek](https://github.com/gumyr/build123d/assets/5196729/b089d6ca-d5d4-41e2-b043-8990b42c65bf)


Another example would be "gluing" blocks side by side. See the following illustration:
![obrazek](https://github.com/gumyr/build123d/assets/5196729/a469a020-3601-4ea0-a0ca-589c72687eb0)


Therefore, I propose a joint type (`OppositeJoint`) allowing two faces to connect without having to adjust or rotate the prepared joints. Joints that connects two sides against. I assume that `x`, `y` of Joint would be in the plane of the "joined" surface, and the `z`-axis would point normal to the surface/object.

Is it more clear?

**Reply by gumyr** [reply]
I see.  How about if there was a `connect_to` parameter `flip` (or something like that) that would apply the `-Location` operation to flip the direction of the joint to allow joints to fit without having to create a redundant joint in the same position?

**Reply by roman-dvorak** [reply]
Something like `flip` would be quite useful. Even better would be a `face-to-face` option, so I wouldn't have to ensure one joint is flipped and the other is not. I support implementing both options (at the same time).

If you agree, I could implement this.

**Reply by gumyr** [reply]
Joint position/orientation isn't defined by a Face but by a Location so `face-to-face` isn't possible.

**Comment by gumyr** [comment]
Having the joints defined within the Part is definitely the way to go.  Users will be able to just use the existing joints and not have to define new ones.
